### ISSUE TITLE
Caffe Training Issue with Caffemodel

### ISSUE THREAD

**Issue by diwgan32** [BVLC/caffe#2721]
Hello All,

I am trying to train a convolutional neural network from scratch to ID some faces. I started out by creating a simple network that has two outputs to distinguish between two identities. When the training was done, the loss seemed to go down to near-zero, so I thought that it had worked. However, the Caffemodel created was over 1GB in size -- which is very unreasonable for the size of the model I'm dealing with. Because of the large file size, when I try to load in the model in python for filter visualization, I get the following error: 

```
[libprotobuf WARNING google/protobuf/io/coded_stream.cc:505] Reading dangerously large protocol message.  If the message turns out to be larger than 1073741824 bytes, parsing will be halted for security reasons.  To increase the limit (or to disable these warnings), see CodedInputStream::SetTotalBytesLimit() in google/protobuf/io/coded_stream.h.
[libprotobuf ERROR google/protobuf/io/coded_stream.cc:171] A protocol message was rejected because it was too big (more than 1073741824 bytes).  To increase the limit (or to disable these warnings), see CodedInputStream::SetTotalBytesLimit() in google/protobuf/io/coded_stream.h.
[libprotobuf WARNING google/protobuf/io/coded_stream.cc:78] The total number of bytes read was 1073741824
F0708 15:15:51.647222  2805 upgrade_proto.cpp:630] Check failed: ReadProtoFromBinaryFile(param_file, param) Failed to parse NetParameter file: ../train/custom_iter_1000.caffemodel

```

Here are is my solver and network definition:

Solver.prototxt

```
net: "custom_network.prototxt"
test_iter: 10
test_interval: 100
base_lr: .001
lr_policy: "step"
gamma: 0.1
stepsize: 300
display: 100
max_iter: 1000
momentum: 0.9
weight_decay: 0.005
snapshot: 1000
snapshot_prefix: "custom"
solver_mode: GPU
```

Network Definition

```
name: "CustomNet"

layers {
  name: "data"
  type: IMAGE_DATA
  image_data_param {
    source: "custom_data_shuffled.txt"
    batch_size: 31;
    new_height: 152
    new_width: 152
	
  }
  transform_param {
	scale: .00390625 
}  
top: "data"
	top: "label"
}
layers {
  name: "conv1"
  type: CONVOLUTION
  blobs_lr: 1
  blobs_lr: 2
  convolution_param {
    num_output: 32
    kernel_size: 11
    stride: 1
    weight_filler {
      type: "gaussian"
      std: .01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
  bottom: "data"
  top: "conv1"
}

layers {
  name: "pool1"
  type: POOLING
  pooling_param {
    pool: MAX
    kernel_size: 3
    stride: 2
  }
  bottom: "conv1"
  top: "pool1"
}
layers {
  name: "conv2"
  blobs_lr: 1
  blobs_lr: 2
  type: CONVOLUTION
  convolution_param {
    num_output: 16
    kernel_size: 9
    weight_filler {
      type: "gaussian"
      std: .01
    }
    bias_filler {
      type: "constant"
	value: 0    
    }
  }
  bottom: "pool1"
  top: "conv2"
}

layers {
  name: "fc6"
  blobs_lr: 1
  blobs_lr: 2
  type: INNER_PRODUCT
  inner_product_param {
    num_output: 4096
    weight_filler {
      type: "gaussian"
      std: .01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
  bottom: "conv2"
  top: "fc6"
}
layers {
  name: "relu6"
  type: RELU
  bottom: "fc6"
  top: "fc6"
}
layers {
  name: "drop6"
  type: DROPOUT
  dropout_param {
    dropout_ratio: 0.5
  }
  bottom: "fc6"
  top: "fc6"
}
layers {
  name: "fc7"
  blobs_lr: 1
  blobs_lr: 2
  type: INNER_PRODUCT
  inner_product_param {
    num_output: 4096
    weight_filler {
      type: "gaussian"
      std: .01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
  bottom: "fc6"
  top: "fc7"
}
layers {
  name: "relu7"
  type: RELU
  bottom: "fc7"
  top: "fc7"
}
layers {
  name: "drop7"
  type: DROPOUT
  dropout_param {
    dropout_ratio: 0.5
  }
  bottom: "fc7"
  top: "fc7"
}
layers {
  name: "fc81"
  blobs_lr: 1
  blobs_lr: 2

  type: INNER_PRODUCT
  inner_product_param {
    weight_filler {
      type: "gaussian"
      std: .01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
    num_output: 2
  }
  bottom: "fc7"
  top: "fc81"
}
layers {
  name: "loss"
  type: SOFTMAX_LOSS
  bottom: "fc81"
  bottom: "label"
  top: "loss"
}
```

Does anyone know why the outputted filesize is so large?

**Comment by rahulbisen** [comment]
Did this get resolved?
I installed caffe(https://github.com/BVLC/caffe/releases/tag/rc2) and trying to follow instructions on deepdream(https://github.com/google/deepdream/blob/master/dream.ipynb)

Getting the same error:

""""""""
I0820 17:37:38.806519 47555 net.cpp:213] Network initialization done.
I0820 17:37:38.806524 47555 net.cpp:214] Memory required for data: 545512320
F0820 17:37:38.807154 47555 upgrade_proto.cpp:935] Check failed: ReadProtoFromBinaryFile(param_file, param) Failed to parse NetParameter file: <path to bvlc_googlenet.caffemodel>
*** Check failure stack trace: ***
""""""""

**Comment by diwgan32** [comment]
Yes, the issue got resolved. It turned out that because the stride of my first convolution layer was 1, it was learning an enormously large number of parameters. I increased the stride, and reduced the size of the input image, and the Caffemodel reduced to a manageable size.

**Comment by lukeyeager** [comment]
Also, you probably don't want the `rc2` version. That was released back in February:
https://github.com/BVLC/caffe/releases

**Comment by rahulbisen** [comment]
The model size doesn't seem to be an issue in my case. I dont get those warnings. 
rc2 is the marked latest released version. And i am using the develop.prototxt and bvlc_googlenet.caffemodel(downloaded by one of the scripts) shipped with rc2 version as well. 
What version / commit you guys suggest to be used??

**Comment by lukeyeager** [comment]
Caffe doesn't really have versions (#486). Everybody just uses the master branch.

**Comment by rahulbisen** [comment]
I got the latest caffe from (https://github.com/BVLC/caffe.git) and tried instructions on deepdream(https://github.com/google/deepdream/blob/master/dream.ipynb)
GoogleNet caffe model: http://dl.caffe.berkeleyvision.org/bvlc_googlenet.caffemodel
Prototxt file: caffe/models//bvlc_googlenet/deploy.prototxt
when I run
"""
net = caffe.Classifier('tmp.prototxt', param_fn,
                       mean = np.float32([104.0, 116.0, 122.0]), # ImageNet mean, training set dependent
                       channel_swap = (2,1,0)) 
"""
which I believe loads the DNN model, i am still getting the same error.

"""
I0821 17:07:53.400810 22824 net.cpp:297] Network initialization done.
I0821 17:07:53.400815 22824 net.cpp:298] Memory required for data: 545512320
F0821 17:07:53.408890 22824 upgrade_proto.cpp:935] Check failed: ReadProtoFromBinaryFile(param_file, param) Failed to parse NetParameter file: /builds/caffe_models/d2f523efb8/share/caffe/models/bvlc_googlenet/bvlc_googlenet.caffemodel
*** Check failure stack trace: ***
Aborted

"""

**Comment by rahulbisen** [comment]
Using released protobuf '2.6.1' from https://github.com/google/protobuf/releases.

**Comment by rahulbisen** [comment]
I am using the cmake setup that is shipped with caffe to compile.
passing options.
"""
cmake -DCMAKE_INSTALL_PREFIX=/tmp/caffe_master_cmake -DBLAS:STRING=MKL -DUSE_CUDNN:BOOL=0 -DBUILD_python:BOOL=1 -DBUILD_matlab:BOOL=0   -DBUILD_SHARED_LIBS:BOOL=1 -DCPU_ONLY:BOOL=1
"""
